pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java

```diff
@@ -3061,6 +3084,104 @@ public void testGlobalTopicPolicies() throws Exception {
 
     }
 
+    @Test
+    public void testInitPolicesCacheAndNotifyListeners() throws Exception {
+        final String topic = testTopic + UUID.randomUUID();
+        admin.topics().createNonPartitionedTopic(topic);
+        pulsarClient.newProducer().topic(topic).create().close();
+
+        // set up policies
+        TopicName topicName = TopicName.get(topic);
+        TopicPolicies localInitPolicy = TopicPolicies.builder().maxConsumerPerTopic(10).build();
+        pulsar.getTopicPoliciesService().updateTopicPoliciesAsync(topicName, localInitPolicy).get();
+        TopicPolicies globalInitPolicy =
+                TopicPolicies.builder().maxConsumerPerTopic(20).maxProducerPerTopic(30).isGlobal(true).build();
+        pulsar.getTopicPoliciesService().updateTopicPoliciesAsync(topicName, globalInitPolicy).get();
+
+        // the policies cache
+        SystemTopicBasedTopicPoliciesService topicPoliciesService
+                = (SystemTopicBasedTopicPoliciesService) pulsar.getTopicPoliciesService();
+
+        // reload namespace to trigger init polices cache and notify listeners
+        admin.namespaces().unload(myNamespace);
+        assertNull(topicPoliciesService.getPoliciesCacheInit(NamespaceName.get(myNamespace)));
+        pulsarClient.newProducer().topic(topic).create().close();
+        Awaitility.await().untilAsserted(
+                () -> assertEquals(topicPoliciesService.getPoliciesCacheInit(NamespaceName.get(myNamespace)).isDone()
+                        && !topicPoliciesService.getPoliciesCacheInit(NamespaceName.get(myNamespace))
+                        .isCompletedExceptionally(), true));
+
+        // the final policies take effect in topic
+        HierarchyTopicPolicies hierarchyTopicPolicies =
+                pulsar.getBrokerService().getTopics().get(topic).get().get().getHierarchyTopicPolicies();
+
+        assertEquals(topicPoliciesService.getTopicPolicies(topicName).getMaxConsumerPerTopic(), 10);
+        assertEquals(topicPoliciesService.getTopicPolicies(topicName, true).getMaxConsumerPerTopic(), 20);
+        assertEquals(hierarchyTopicPolicies.getMaxConsumerPerTopic().getGlobalTopicValue(), 20);
+        assertEquals(hierarchyTopicPolicies.getMaxConsumerPerTopic().getLocalTopicValue(), 10);
+        assertEquals(hierarchyTopicPolicies.getMaxConsumerPerTopic().get(), 10);
+
+        assertEquals(topicPoliciesService.getTopicPolicies(topicName).getMaxProducerPerTopic(), null);
+        assertEquals(topicPoliciesService.getTopicPolicies(topicName, true).getMaxProducerPerTopic(), 30);
+        assertEquals(hierarchyTopicPolicies.getMaxProducersPerTopic().getGlobalTopicValue(), 30);
+        assertEquals(hierarchyTopicPolicies.getMaxProducersPerTopic().getLocalTopicValue(), null);
+        assertEquals(hierarchyTopicPolicies.getMaxProducersPerTopic().get(), 30);
+    }
+
+    @Test
+    public void testInitPolicesCacheAndNotifyListenersAfterCompaction() throws Exception {
```

Comment:
A new channel to trace the comment https://github.com/apache/pulsar/pull/21212#pullrequestreview-1636840630

> @mattisonchao 
> Could you please add a test to verify the changes and explain how you can avoid the compactor to override local policy with global policy(same event key) :) 


Agree with @mattisonchao , the test flow might be like this:
- set local policies
- set global policies
- wait for the compaction task to finish
- read the `change_event` topic to verify there are two events for this topic. Or find a way to start a new `SystemTopicBasedTopicPoliciesService` and verify the two caches (`policiesCache` and `globalPoliciesCache`) each has one item for this topic

Reply:
I think the test case is exactly what you described, please correct me if I have misunderstood.

Reply:
I do not think the test you provided can cover the case. Could you try to run the test below? By the way, 

```java
private void triggerAndWaitNewTopicCompaction(String topicName) throws Exception {
    PersistentTopic tp =
            (PersistentTopic) pulsar.getBrokerService().getTopic(topicName, false).join().get();
    // Wait for the old task finish.
    Awaitility.await().untilAsserted(() -> {
        CompletableFuture<Long> compactionTask = WhiteboxImpl.getInternalState(tp, "currentCompaction");
        assertTrue(compactionTask == null || compactionTask.isDone());
    });
    // Trigger a new task.
    tp.triggerCompaction();
    // Wait for the new task finish.
    Awaitility.await().untilAsserted(() -> {
        CompletableFuture<Long> compactionTask = WhiteboxImpl.getInternalState(tp, "currentCompaction");
        assertTrue(compactionTask == null || compactionTask.isDone());
    });
}

/***
 * It is not a thread safety method, something will go to a wrong pointer if there is a task is trying to load a
 * topic policies.
 */
private void clearTopicPoliciesCache() {
    TopicPoliciesService topicPoliciesService = pulsar.getTopicPoliciesService();
    if (topicPoliciesService instanceof TopicPoliciesService.TopicPoliciesServiceDisabled) {
        return;
    }
    assertTrue(topicPoliciesService instanceof SystemTopicBasedTopicPoliciesService);

    Map<NamespaceName, CompletableFuture<Void>> policyCacheInitMap =
            WhiteboxImpl.getInternalState(topicPoliciesService, "policyCacheInitMap");
    for (CompletableFuture<Void> future : policyCacheInitMap.values()) {
        future.join();
    }
    Map<TopicName, TopicPolicies> policiesCache =
            WhiteboxImpl.getInternalState(topicPoliciesService, "policiesCache");
    Map<TopicName, TopicPolicies> globalPoliciesCache =
            WhiteboxImpl.getInternalState(topicPoliciesService, "globalPoliciesCache");

    policyCacheInitMap.clear();
    policiesCache.clear();
    globalPoliciesCache.clear();
}

@Test
public void testGetTopicPoliciesWhenDeleteTopicPolicy2() throws Exception {
    final String tpName = BrokerTestUtil.newUniqueName("persistent://" + myNamespace + "/tp");
    final String tpNameChangeEvents = "persistent://" + myNamespace + "/" + NAMESPACE_EVENTS_LOCAL_NAME;
    final String subscriptionName = "s1";
    final int rateMsgLocal = 2000;
    final int rateMsgGlobal = 1000;
    admin.topics().createNonPartitionedTopic(tpName);
    admin.topics().createSubscription(tpName, subscriptionName, MessageId.earliest);

    // Set global policy and local policy.
    DispatchRate dispatchRateLocal = new DispatchRateImpl(rateMsgLocal, 1, false, 1);
    DispatchRate dispatchRateGlobal = new DispatchRateImpl(rateMsgGlobal, 1, false, 1);
    admin.topicPolicies(false).setDispatchRate(tpName, dispatchRateLocal);
    admin.topicPolicies(true).setDispatchRate(tpName, dispatchRateGlobal);

    // Trigger __change_events compaction and clear topic policies cache.
    triggerAndWaitNewTopicCompaction(tpNameChangeEvents);
    clearTopicPoliciesCache();

    // Reload the topic policies.
    // Verify the local policies was affected.
    Optional<TopicPolicies> topicPoliciesOptional =
            pulsar.getTopicPoliciesService().getTopicPoliciesAsync(TopicName.get(tpName)).join();
    assertTrue(topicPoliciesOptional.isPresent());
    assertEquals(topicPoliciesOptional.get().getDispatchRate().getDispatchThrottlingRateInMsg(),
            rateMsgLocal);

    // cleanup.
    admin.topics().delete(tpName, false);
}
```

Reply:
I think the test case(testInitPolicesCacheAndNotifyListenersAfterCompaction) can cover the fix about compaction, the test case will fail without this fix. 
I also added the test cases you provided above, and it's passed.

Reply:
@chenhongSZ 

Sorry, there is a mistake in the test. Could you change the line like below? Thanks

```java
private void triggerAndWaitNewTopicCompaction(String topicName) throws Exception {
    PersistentTopic tp =
            (PersistentTopic) pulsar.getBrokerService().getTopic(topicName, false).join().get();
    // Wait for the old task finish.
    Awaitility.await().untilAsserted(() -> {
        CompletableFuture<Long> compactionTask = WhiteboxImpl.getInternalState(tp, "currentCompaction");
        assertTrue(compactionTask == null || compactionTask.isDone());
    });
    // Trigger a new task.
    tp.triggerCompaction();
    // Wait for the new task finish.
    Awaitility.await().untilAsserted(() -> {
        CompletableFuture<Long> compactionTask = WhiteboxImpl.getInternalState(tp, "currentCompaction");
        assertTrue(compactionTask == null || compactionTask.isDone());
    });
}

@Test
public void testLocalPolicyAffectAfterCompaction() throws Exception {
    final String tpName = BrokerTestUtil.newUniqueName("persistent://" + myNamespace + "/tp");
    final String tpNameChangeEvents = "persistent://" + myNamespace + "/" + NAMESPACE_EVENTS_LOCAL_NAME;
    final String subscriptionName = "s1";
    final int rateMsgLocal = 2000;
    final int rateMsgGlobal = 1000;
    admin.topics().createNonPartitionedTopic(tpName);
    admin.topics().createSubscription(tpName, subscriptionName, MessageId.earliest);

    // Set global policy and local policy.
    DispatchRate dispatchRateLocal = new DispatchRateImpl(rateMsgLocal, 1, false, 1);
    DispatchRate dispatchRateGlobal = new DispatchRateImpl(rateMsgGlobal, 1, false, 1);
    admin.topicPolicies(false).setDispatchRate(tpName, dispatchRateLocal);
    admin.topicPolicies(true).setDispatchRate(tpName, dispatchRateGlobal);

    // Trigger __change_events compaction.
    triggerAndWaitNewTopicCompaction(tpNameChangeEvents);

    // Create a new SystemTopicBasedTopicPoliciesService and verify the local policies was affected.
    Optional<TopicPolicies> topicPoliciesOptional =
            new SystemTopicBasedTopicPoliciesService(pulsar).getTopicPoliciesAsync(TopicName.get(tpName)).join();
    assertTrue(topicPoliciesOptional.isPresent());
    assertEquals(topicPoliciesOptional.get().getDispatchRate().getDispatchThrottlingRateInMsg(),
            rateMsgLocal);

    // cleanup.
    admin.topics().delete(tpName, false);
}
```

Reply:
The changes below solved this issue. 

https://github.com/apache/pulsar/pull/21212/files#diff-9d2948d863c111e4be6d508a1c573667a1326b98c4314e917ba9e344bb61dc27R665-L635

<img width="970" alt="截屏2023-10-09 00 11 31" src="https://github.com/apache/pulsar/assets/25195800/eb5525bf-505d-4589-85e2-85c554cf7dd4">

@chenhongSZ 

But this fix also changed the rule of key building, which makes it not forward-compatible. For example:
- The user updates local policies, and Pulsar sends a message with the key "persistent://public/default/topic_1"
- The user upgrades the Pulsar cluster to the newest version.
- The user updates local policies, and Pulsar sends a message with the key "persistent/public/default/topic_1/false"
- The user updates global policies, and Pulsar sends a message with the key "persistent/public/default/topic_1/true"

At this time, there should be three messages in the system topic, but the behavior will be correct even if it is not expected, but the behavior will be wrong after the steps below.

- Delete the topic
  - Pulsar deletes the key "persistent/public/default/topic_1/false"
  - Pulsar deletes the key "persistent/public/default/topic_1/true"
- Create a new topic with the same name as the old one.

Currently, the message whose key is "persistent://public/default/topic_1" is still there.

I think once the new key is generated, the data of the old key should be deleted. And could you also add a test for this case?

By the way, the original rule is `persistent://xxx`, do not change it to `persistent/xxx`.

Reply:
Please negotiate this special key name. Do not make it casually. :)

Reply:
I will fix the compatibility problem and add a test later.
 I didn't change the rules of the key building, just added a property, because we need to get the topic name from the key
original: `persistent://tenant/namespace/topic/`
new:`persistent://tenant/namespace/topic/isGlobal`

Reply:
> @chenhongSZ
> 
> Sorry, there is a mistake in the test. Could you change the line like below? Thanks
> 
> ```java
> private void triggerAndWaitNewTopicCompaction(String topicName) throws Exception {
>     PersistentTopic tp =
>             (PersistentTopic) pulsar.getBrokerService().getTopic(topicName, false).join().get();
>     // Wait for the old task finish.
>     Awaitility.await().untilAsserted(() -> {
>         CompletableFuture<Long> compactionTask = WhiteboxImpl.getInternalState(tp, "currentCompaction");
>         assertTrue(compactionTask == null || compactionTask.isDone());
>     });
>     // Trigger a new task.
>     tp.triggerCompaction();
>     // Wait for the new task finish.
>     Awaitility.await().untilAsserted(() -> {
>         CompletableFuture<Long> compactionTask = WhiteboxImpl.getInternalState(tp, "currentCompaction");
>         assertTrue(compactionTask == null || compactionTask.isDone());
>     });
> }
> 
> @Test
> public void testLocalPolicyAffectAfterCompaction() throws Exception {
>     final String tpName = BrokerTestUtil.newUniqueName("persistent://" + myNamespace + "/tp");
>     final String tpNameChangeEvents = "persistent://" + myNamespace + "/" + NAMESPACE_EVENTS_LOCAL_NAME;
>     final String subscriptionName = "s1";
>     final int rateMsgLocal = 2000;
>     final int rateMsgGlobal = 1000;
>     admin.topics().createNonPartitionedTopic(tpName);
>     admin.topics().createSubscription(tpName, subscriptionName, MessageId.earliest);
> 
>     // Set global policy and local policy.
>     DispatchRate dispatchRateLocal = new DispatchRateImpl(rateMsgLocal, 1, false, 1);
>     DispatchRate dispatchRateGlobal = new DispatchRateImpl(rateMsgGlobal, 1, false, 1);
>     admin.topicPolicies(false).setDispatchRate(tpName, dispatchRateLocal);
>     admin.topicPolicies(true).setDispatchRate(tpName, dispatchRateGlobal);
> 
>     // Trigger __change_events compaction.
>     triggerAndWaitNewTopicCompaction(tpNameChangeEvents);
> 
>     // Create a new SystemTopicBasedTopicPoliciesService and verify the local policies was affected.
>     Optional<TopicPolicies> topicPoliciesOptional =
>             new SystemTopicBasedTopicPoliciesService(pulsar).getTopicPoliciesAsync(TopicName.get(tpName)).join();
>     assertTrue(topicPoliciesOptional.isPresent());
>     assertEquals(topicPoliciesOptional.get().getDispatchRate().getDispatchThrottlingRateInMsg(),
>             rateMsgLocal);
> 
>     // cleanup.
>     admin.topics().delete(tpName, false);
> }
> ```

fixed

Reply:
I've fixed the compatibility problem and added a test to cover this change. PTAL